notification_api.cpp

```diff
@@ -73,9 +91,123 @@ void NotificationApi::startServer() {
 void NotificationApi::stopServer() {
 	m_server->stop();
 }
+
 /**
  * Wait for the HTTP server to shutdown
  */
 void NotificationApi::wait() {
 	m_thread->join();
 }
+
+/**
+ * Initialise the API entry points for the common data resource and
+ * the readings resource.
+ */
+void NotificationApi::initResources()
+{       
+	m_server->resource[RECEIVE_NOTIFICATION]["POST"] = notificationReceiveWrapper;
+}
+
+/**
+ * Handle a exception by sendign back an internal error
+ */
+void NotificationApi::internalError(shared_ptr<HttpServer::Response> response,
+				    const exception& ex)
+{
+	string payload = "{ \"Exception\" : \"";
+
+	payload = payload + string(ex.what());
+	payload = payload + "\"";
+
+	Logger *logger = Logger::getLogger();
+	logger->error("NotificationApi Internal Error: %s\n", ex.what());
+
+	this->respond(response,
+		      SimpleWeb::StatusCode::server_error_internal_server_error,
+		      payload);
+}
+
+/**
+ * Construct an HTTP response with the 200 OK return code using the payload
+ * provided.
+ *
+ * @param response	The response stream to send the response on
+ * @param payload	The payload to send
+ */
+void NotificationApi::respond(shared_ptr<HttpServer::Response> response,
+			      const string& payload)
+{
+	*response << "HTTP/1.1 200 OK\r\nContent-Length: "
+		  << payload.length() << "\r\n"
+		  <<  "Content-type: application/json\r\n\r\n" << payload;
+}
+
+/**
+ * Construct an HTTP response with the specified return code using the payload
+ * provided.
+ *
+ * @param response	The response stream to send the response on
+ * @param code		The HTTP esponse code to send
+ * @param payload	The payload to send
+ */
+void NotificationApi::respond(shared_ptr<HttpServer::Response> response,
+			      SimpleWeb::StatusCode code,
+			      const string& payload)
+{
+	*response << "HTTP/1.1 " << status_code(code) << "\r\nContent-Length: "
+		  << payload.length() << "\r\n"
+		  <<  "Content-type: application/json\r\n\r\n" << payload;
+}
+
+/**
+ * Add data provided in the payload of callback API call
+ * into the notification queue.
+ *
```

Comment:
Probably worth documenting what this callback is about for future readers.

I.e. This is called by the storage service when new data arrives for an asset in which we have registered an interest.